File: src/predict_binomial.f90
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
!!! Fortran routines for calculation of binomial class labels
!!!
!!! @author: Simon Dirmeier
!!! @email: simon.dirmeier@gmx.de
!!!

!! compute the binomial class labels
      subroutine predict_binomial(na, ntrain, nnew, K, ctrain, sigtrain, xnew,& 
                                 trainidx, newidx, mnew, ps, pm, Kn)
! load modules
      use gpr_approx
      use gpr_sampler
      use gpr_linalg
      implicit none
      integer, intent(in)           :: na, ntrain, nnew
      integer, intent(in)           :: trainidx(ntrain), newidx(nnew)
      double precision, intent(in)  :: ctrain(ntrain), K(na, na) 
      double precision, intent(in)  :: sigtrain(ntrain), xnew(nnew)
      double precision, intent(out) :: mnew(nnew), ps(nnew), pm(nnew), Kn(nnew, nnew)
! worker arrays
      double precision :: knn, me, pre, var, m, rn
      double precision, allocatable :: cm(:), Ktt(:, :), D(:, :), Knt(:), Cu(:, :)
      integer :: i, cidx, nnewidxst
!
      allocate(cm(ntrain), Ktt(ntrain, ntrain), D(ntrain, ntrain))
      allocate(Knt(ntrain), Cu(ntrain, ntrain))
!
      Ktt = K(1:ntrain, 1:ntrain)
! compute log probability of the class mapping of c.train given
! the posterior mean
! (see references: Barber p406, Equation~19.5.24;
!                  Rasmussen p44, Equation~3.15 & Equation~3.21)
      cm = ctrain - sigtrain
      D = 0.0
! worker       
      Cu  = Ktt + D
      call solve(ntrain, Cu)
! compute the inverse of the Hessian of the log-likelihood      
      forall(i = 1:ntrain) D(i, i) = sigtrain(i) * (1 - sigtrain(i))
      do i = 1, nnew
          cidx = newidx(i)
! get the covariance of the training points and a new input          
          Knt = K(1:ntrain, cidx)
! get the variance of the new input point          
          knn = K(cidx, cidx)
! compute the posterior mean of the predictive distribution
! (see references: Barber p406, Equation~19.5.24;
!                  Rasmussen p44, Equation~3.21)          
          m   = DOT_PRODUCT(Knt, cm)
! compute the posterior variance of the predictive distribution
! (see references: Barber p406, Equation~19.5.26;
!                  Rasmussen p44, Equation~3.24)          
          var = knn - DOT_PRODUCT(MATMUL(Knt, Cu), Knt)
! compute the mean class probability mapping
! (see references: Barber p406, Equation~19.5.27;
!                  Rasmussen p44, Equation~3.25)          
          me  = lognint(m, var)
          rn = rnorm(m, var)
! compute a sample from the posterior class mapping          
          pre =  1 / (1 + exp(-1 * rn ))
          ps(i) = pre
          pm(i) = me
      end do
      nnewidxst = ntrain + 1
! mean of posterior       
      mnew = MATMUL(K(nnewidxst:na, 1:ntrain), cm)
! variance of posterior      
      Kn = K(newidx, newidx) / MATMUL(MATMUL(K(nnewidxst:na, 1:ntrain), Cu), &
                                      K(1:ntrain, nnewidxst:na))
!
      deallocate(cm, Ktt, Knt, D, Cu)
      return
      end subroutine

File: R/predict_binomial.R
#' @title Predict a new set of data points x.new using Gaussian process classification
#'
#' @noRd
#' @param object  an \code{lvgpc.data} object
#' @return an \code{lvgpc.pred} object
#' \item{c.predict }{ the predicted c* values given the \code{lvgpc.data} object}
#' \item{c.labels }{ the predicted c* labale given the \code{lvgpc.data} object}
#' \item{mean.c.predict }{ the predicted mean c* values given the \code{lvgpc.data} object}
#' \item{mean }{ the (approximated) posterior mean values}
#' \item{cov  }{ the (approximated) posterior covariance/kernel}
#' @references
#'  Rasmussen C.E. and Williams C.K.I. (2006), \emph{Gaussian Processes for Machine Learning}, MIT Press \cr
#'  \url{http://www.gaussianprocess.org/gpml/} \cr \cr
#'  Barber D. (2015), \emph{Bayesian Reasoning and Machine Learning}, Cambridge University Press \cr
#'  \url{http://web4.cs.ucl.ac.uk/staff/D.Barber/pmwiki/pmwiki.php?n=Brml.Online}
predict.binomial <- function(obj, ...)
{
  x.train <- obj@x.train
  c.train <- obj@c.train
  x.new   <- obj@x.new
  n <- length(x.train)
  train <- 1:n
  test <- 1:length(x.new) + n
  # approximate the mode/mean of the posterior, its covariance,
  #  the estimated class mappings and the Hessian of the log-likelihood
  approx <- approx.posterior(obj, ...)
  # logistic transform of posterior mean values
  sig <- approx$log.transform
  # joint covariance matrix of x.train and x.new
  K <- approx$cov
  res <- .predict.binomial(K, c.train, sig, x.new, train, test)
  c.labels <- rep(0, length(res$pred.samples))
  c.labels[res$pred.samples >= 0.5] <- 1
  ret <- base::list(mean.c.predict=res$pred.means,
                    c.predict=res$pred.samples,
                    c.labels = c.labels,
                    cov=res$cov.new.new,
                    mean=res$m.new)
  ret$call <- match.call()
  ret
}

#' @noRd
.predict.binomial <- function(K, c.train, sig, x.new, train, test)
{
  ntrain <- length(train)
  nnew   <- length(test)
  n      <- ntrain + nnew
  me <- pred.sample <- pred.means <- rep(0, nnew)
  K.new.new <- matrix(0, nnew, nnew)
  res <- .Fortran("predict_binomial",
                  na=as.integer(n),
                  ntrain=as.integer(ntrain),
                  nnew=as.integer(nnew),
                  K=K,
                  ctrain=c.train,
                  sigtrain=sig,
                  xnew=x.new,
                  trainidx=as.integer(train),
                  newidx=as.integer(test),
                  mnew=me,
                  ps=pred.sample,
                  pm=pred.means,
                  Kn=K.new.new,
                  PACKAGE="gpR")
  list(m.new=res$mnew,
       cov.new.new=res$Kn,
       pred.samples=res$ps,
       pred.means=res$pm)
}
